backends/backend_openssl3.c

```diff
@@ -2472,3 +2472,629 @@ static void openssl_ecdsa_backend(void)
+ ************************************************/
+@@ -2569,3 +2692,498 @@ static void openssl_ecdsa_backend(void)
+{
+	register_ecdsa_impl(&openssl_ecdsa);
+}
+
+
+/************************************************
+ * RSA interface functions
+ ************************************************/
+
+#define RSA 		EVP_PKEY
+#define RSA_free(a)		EVP_PKEY_free(a)
+#define EVP_DigestSignInit(a,b,c,d,e)		EVP_DigestSignInit_ex(a,b,EVP_MD_name(c),NULL,NULL,e,NULL)
+#define EVP_DigestVerifyInit(a,b,c,d,e)		EVP_DigestVerifyInit_ex(a,b,EVP_MD_name(c),NULL,NULL,e,NULL)
+
+static int openssl_rsa_keygen_prime(struct rsa_keygen_prime_data *data,
+					flags_t parsed_flags)
+{
+	BIGNUM *e = NULL, *p = NULL, *q = NULL;
+	EVP_PKEY_CTX *ctx = NULL;
+	EVP_PKEY *pkey = NULL;
+	OSSL_PARAM_BLD *bld = NULL;
+	OSSL_PARAM *params = NULL;
+	int ret = 0;
+
+	(void)parsed_flags;
+
+	if (!data->e.len) {
+		logger(LOGGER_WARN, "RSA E missing\n");
+		return -EINVAL;
+	}
+
+	bld = OSSL_PARAM_BLD_new();
+	logger_binary(LOGGER_DEBUG, data->e.buf, data->e.len, "e");
+	logger_binary(LOGGER_DEBUG, data->p.buf, data->p.len, "p");
+	logger_binary(LOGGER_DEBUG, data->q.buf, data->q.len, "q");
+
+	e = BN_bin2bn((const unsigned char *) data->e.buf, (int)data->e.len, e);
+	CKNULL(e, -ENOMEM);
+
+	p = BN_bin2bn((const unsigned char *) data->p.buf, (int)data->p.len, p);
+	CKNULL(p, -ENOMEM);
+	if (BN_is_zero(p))
+		BN_one(p);
+	q = BN_bin2bn((const unsigned char *)data->q.buf, (int)data->q.len, q);
+	CKNULL(q, -ENOMEM);
+	OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, e);
+	if (p != NULL) {
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR1, p);
+	}
+	if (q != NULL) {
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR2, q);
+	}
+	params = OSSL_PARAM_BLD_to_param(bld);
+	ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", NULL);
+	EVP_PKEY_keygen_init(ctx);
+	ret = EVP_PKEY_CTX_set_params(ctx, params);
+	data->keygen_success = 1;
+
+	if(EVP_PKEY_generate(ctx, &pkey) <= 0)
+		data->keygen_success = 0;
+	if(BN_check_prime(p, NULL, NULL) <= 0)
+		data->keygen_success = 0;
+	if(BN_check_prime(q, NULL, NULL) <= 0)
+		data->keygen_success = 0;
+
+out:
+	if(params)
+		OSSL_PARAM_free(params);
+	if(bld)
+		OSSL_PARAM_BLD_free(bld);
+	if(ctx)
+		EVP_PKEY_CTX_free(ctx);
+	return ret;
+}
+
+int openssl_rsa_keygen_internal(struct buffer *ebuf, uint32_t modulus,
+					EVP_PKEY **outkey, struct buffer *nbuf,
+					struct buffer *dbuf, struct buffer *pbuf,
+					struct buffer *qbuf)
+{
+	BIGNUM *e = NULL;
+	EVP_PKEY_CTX *ctx = NULL;
+	OSSL_PARAM_BLD *bld = NULL;
+	EVP_PKEY *pkey = NULL;
+	size_t p_len = 0, q_len=0, egen_len=0, d_len=0,n_len=0;
+	unsigned int retry = 0;
+	int ret = 0;
+	unsigned char *n , *p , *q , *d,*egen;
+
+	if (!ebuf->len) {
+		unsigned int a;
+		uint8_t bitsset = 0;
+
+		/* WARNING Buffer must be at least 3 bytes in size ! */
+		CKINT(alloc_buf(sizeof(unsigned int), ebuf));
+
+		/* generate random odd e */
+		RAND_bytes(ebuf->buf, (int)ebuf->len);
```

Comment:
is there a requirement for the entropy to be fips compliant ?

Reply:
No, there is no such requirement as the test result for rsa are already verified by lab.